zipextimporter.py

```diff
@@ -42,137 +42,196 @@
+        if not mi.is_ext:
+            if hasattr(zipimporter, "create_module"):
+                return super().create_module(spec)
             else:
-                suffixes = self._suffixes
-            for s in suffixes:
-                path = filename + s
-                if path in self._files:
-                    if verbose > 1:
-                        sys.stderr.write("# found %s in zipfile %s\n"
-                                         % (path, self.archive))
-                    spec = util.find_spec(fullname, path)
-                    mod = _memimporter.import_module(fullname, path,
-                                                     initname,
-                                                     self.get_data, spec)
-                    mod.__file__ = "%s\\%s" % (self.archive, path)
-                    mod.__loader__ = self
-                    mod.__memimported__ = True
-                    sys.modules[fullname] = mod
-                    if verbose:
-                        sys.stderr.write("import %s # loaded from zipfile %s\n"
-                                         % (fullname, mod.__file__))
-                    return mod
-            raise zipimport.ZipImportError("can't find module %s" % fullname) from err
-
-    if sys.version_info >= (3, 10):
-        def create_module(self, spec):
-            mod =  super().create_module(spec)
-            if mod is None:
-                verbose = _memimporter.get_verbose_flag()
-                fullname = spec.name
-
-                filename = fullname.replace(".", "\\")
-                suffixes = self._suffixes
-                initname = "PyInit_" + fullname.split(".")[-1]
-
-                for s in suffixes:
-                    path = filename + s
-                    if path in self._files:
-                        if verbose > 1:
-                            sys.stderr.write("# found %s in zipfile %s\n"
-                                            % (path, self.archive))
-                        mod = _memimporter.import_module(fullname, path,
-                                                        initname,
-                                                        self.get_data, spec)
-                        mod.__file__ = "%s\\%s" % (self.archive, path)
-                        mod.__loader__ = self
-                        mod.__memimported__ = True
-                        if verbose:
-                            sys.stderr.write("import %s # loaded from zipfile %s\n"
-                                            % (fullname, mod.__file__))
-                        return mod
-                # raise zipimport.ZipImportError("can't find module %s" % fullname)
-
-    if sys.version_info >= (3, 10):
-        def exec_module(self, module):
-            if hasattr(module, '__memimported__'):
+                return super().load_module(fullname)
+
+        spec._set_fileattr = True  # has_location, use for reload
```

Comment:
Should `spec.has_location` be used as better?

Reply:
In principle, yes, but we need to make sure this passes the tests.

Reply:
Yes, but I forgot why test fail, seems issue is not related this PR. I will push new commit in later time.